[MRG] Sliced wasserstein

AdrienCorenflos opened the pull request:
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## Motivation and context / Related issue
Implement SWD: https://github.com/PythonOT/POT/issues/202


## How has this been tested (if it applies)
Added specific tests (positive definiteness + matching the EMD in the 1D case)


## Checklist

- [X] The documentation is up-to-date with the changes I made.
- [X] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] All tests passed, and additional code has been covered with new tests.

Not sure why yet but the stuff doesn't build.

I'm publishing this as a draft as I have some other changes in my branch that are pending for another merge (cf this: https://github.com/PythonOT/POT/issues/200)

AdrienCorenflos commented:
It doesn't build because of the string formatting: f"stuff" which is only supported from python 3.6. You guys sure you want to support python 3.5?

rflamary commented:
yes we keep it, we can have this discussion for the next major release but for now we still support 3.5.

Note that you have a lot of things in this PR that should not be here (and in other PR):
+ Change for sinkhorn in log (i agree for a new solver but it should be a new one, you do not change the current default)
+ Correction for the doc of barycenter (already in another PR)

rflamary commented:
Sorry did not see the revert up there about log ;)

AdrienCorenflos commented:
I reverted a bunch of commits in this branch, you should compare the files,
not look at the history (sinkhorn in log is not there anymore).
Correct for the docstring. I was sloppy with my branches, so I'm just
waiting for the doc one to be merged so the branch will be up to date.

Le lun. 20 juil. 2020 à 14:56, Rémi Flamary <notifications@github.com> a
écrit :

> yes we keep it, we can have this discussion for the next major release but
> for now we still support 3.5.
>
> Note that you have a lot of things in this PR that should not be here (and
> in other PR):
>
>    - Change for sinkhorn in log (i agree for a new solver but it should
>    be a new one, you do not change the current default)
>    - Correction for the doc of barycenter (already in another PR)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PythonOT/POT/pull/203#issuecomment-660982034>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYGFZ5X2FXQOFVU3C4L5JDR4QWFDANCNFSM4PCHDJXA>
> .
>

rflamary commented:
done and updated you PR

AdrienCorenflos commented:
Cool I corrected the string formatting for py3.5 support.
Also honestly, all this SWD should probably be done directly in cython as
the calculation can be vectorized, but I guess that can be left for a
hypothetical future...

Le lun. 20 juil. 2020 à 15:03, Rémi Flamary <notifications@github.com> a
écrit :

> done and updated you PR
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PythonOT/POT/pull/203#issuecomment-660985258>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYGFZ3OEVACFKPF5Q3DAWLR4QXC3ANCNFSM4PCHDJXA>
> .
>

AdrienCorenflos commented:
Meant parallelized, not vectorized.

Le lun. 20 juil. 2020 à 15:05, Adrien Corenflos <adrien.corenflos@gmail.com>
a écrit :

> Cool I corrected the string formatting for py3.5 support.
> Also honestly, all this SWD should probably be done directly in cython as
> the calculation can be vectorized, but I guess that can be left for a
> hypothetical future...
>
> Le lun. 20 juil. 2020 à 15:03, Rémi Flamary <notifications@github.com> a
> écrit :
>
>> done and updated you PR
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/PythonOT/POT/pull/203#issuecomment-660985258>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEYGFZ3OEVACFKPF5Q3DAWLR4QXC3ANCNFSM4PCHDJXA>
>> .
>>
>

AdrienCorenflos commented:
Cool, looks like it's building just fine

rflamary commented on ot/__init__.py:
```diff
@@ -39,6 +39,7 @@
 from .bregman import sinkhorn, sinkhorn2, barycenter
 from .unbalanced import sinkhorn_unbalanced, barycenter_unbalanced, sinkhorn_unbalanced2
 from .da import sinkhorn_lpl1_mm
+from .sliced import sliced
```
you should also add sliced in `__all__`

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,100 @@
+import numpy as np
+
+
+def _random_projections(n_projections, dimension, random_state):
```
`get_random_projection` ? 

Also this function can be used elsewhere I would not hide it and it needs a more detailled doc

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,100 @@
+    return projections
+
+
+def sliced(X_s, X_t, a=None, b=None, n_projections=50, seed=None):
```
all parameter `log=False` that will allow returning information (at least the random projection operator)

rflamary replied:
Also you should name the function `sliced_wasserstein_distance`

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,100 @@
+
+    res = 0.
+    for projection in projections:
+        X_s_proj = X_s @ projection
```
please use dot. I know it works on all recent Python 3 but we use .dot everywhere else in POT and I would like to keep it uniform.

Also you should project before for all direction using `X_s.dot(projections)` that can be accelerated by blas instead of doing it in the loop.

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,100 @@
+        X_s_proj = X_s @ projection
+        X_t_proj = X_t @ projection
+        res += emd2_1d(X_s_proj, X_t_proj, a, b, log=False, dense=False)
+    return (res / n_projections) ** 0.5
```
pleas also return a log dictionary containing the loss for each directions and the projections if `log=True`

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,100 @@
+    References
+    ----------
+
+    .. [1] S. Kolouri et al., Generalized Sliced Wasserstein Distances, Advances in Neural Information Processing Systems (NIPS) 33, 2019
```
Please cite the following contribution that is much earlier : 
Bonneel, Nicolas, et al. "Sliced and radon wasserstein barycenters of measures." Journal of Mathematical Imaging and Vision 51.1 (2015): 22-45.

you should also add it to the readme and use the same number as in the readme for reference.

AdrienCorenflos commented:
Thanks for the thorough review. I'll amend all this early next week.

Adrien

On Fri, 31 Jul 2020, 17:01 Rémi Flamary, <notifications@github.com> wrote:

> *@rflamary* commented on this pull request.
> ------------------------------
>
> In ot/sliced.py
> <https://github.com/PythonOT/POT/pull/203#discussion_r463629013>:
>
> > +#
> +# License: MIT License
> +
> +
> +import numpy as np
> +
> +
> +def _random_projections(n_projections, dimension, random_state):
> +    """Samples n_projections times dimension normal distributions"""
> +    projections = random_state.normal(0., 1., [n_projections, dimension])
> +    norm = np.linalg.norm(projections, ord=2, axis=1, keepdims=True)
> +    projections = projections / norm
> +    return projections
> +
> +
> +def sliced(X_s, X_t, a=None, b=None, n_projections=50, seed=None):
>
> Also you should name the function sliced_wasserstein_distance
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PythonOT/POT/pull/203#discussion_r463629013>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEYGFZ675GRJHRM5DFVPEKTR6LFCTANCNFSM4PCHDJXA>
> .
>

AdrienCorenflos commented:
Feels like it should be good @rflamary

AdrienCorenflos commented:
@rflamary I think I included all your comments, do you see anything left?

rflamary commented:
Sorry this is an hectic time on my end, I see some small comments but don't have time for a proper code review yet. Will do it asap. Only minor things remain thank you for taking into account all my comments.

AdrienCorenflos commented:
Btw on the back of this I (very) can easily do the (heuristical max slice), is that something that would be of interest?

rflamary commented on README.md:
```diff
@@ -263,3 +265,7 @@ You can also post bug reports and feature requests in Github issues. Make sure t
+
+[31] Bonneel, Nicolas, et al. "Sliced and radon wasserstein barycenters of measures." Journal of Mathematical Imaging and Vision 51.1 (2015): 22-45
+
+[32] S. Kolouri et al., Generalized Sliced Wasserstein Distances, Advances in Neural Information Processing Systems (NIPS) 33, 2019
```
We do not implement this method yet so it should not be in the readme

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,144 @@
+    r"""
+    Computes a Monte-Carlo approximation of the 2-Sliced Wasserstein distance
+    .. math::
+        \mathcal{SWD}_2(\mu, \nu) = \underset{\theta \sim \mathcal{U}(\mathbb{S}^{d-1})}{\mathbb{E}}[\mathcal{W}_2^2(\theta_\# \mu, \theta_\# \nu)]^{\frac{1}{2}}
```
the math environment fails in the doc see:
https://331-71472695-gh.circle-artifacts.com/0/dev/gen_modules/ot.sliced.html#ot.sliced.sliced_wasserstein_distance

maybe a line return missing?

AdrienCorenflos replied:
Correct. Just checked it, doc is ok now

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,144 @@
+    ----------
+
+    .. [31] Bonneel, Nicolas, et al. "Sliced and radon wasserstein barycenters of measures." Journal of Mathematical Imaging and Vision 51.1 (2015): 22-45
+    .. [32] S. Kolouri et al., Generalized Sliced Wasserstein Distances, Advances in Neural Information Processing Systems (NIPS) 33, 2019
```
Again this method is not implemented here. we can add it when we acually ilmplement it ;)

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,144 @@
+    for X_s_proj, X_t_proj in zip(X_s_projections, X_t_projections):
+        emd = emd2_1d(X_s_proj, X_t_proj, a, b, log=False, dense=False)
+        if projected_emd is not None:
+            projected_emd.append(emd)
```
Sorry to complain again but you should not use a list with append to store a vector of double of known length, 

Please initialize `projected_emd` with zeros and set its values in the loop (either use a range in the loop or an enumerate in addition to the zip)

rflamary commented on ot/sliced.py:
```diff
@@ -0,0 +1,144 @@
+
+    res = (res / n_projections) ** 0.5
+    if log:
+        return res, {"projections": projections.tolist(), "projected_emds": projected_emd}
```
why lists? the function should return numpy array

AdrienCorenflos replied:
No reason, I looked at the bregman one to the preferred format. Didn't think that the "err" item of the log dict was a list because we don't know it's size a priori.

rflamary reviewed (commented):
Here are some remaining comments on the code.

yes it would be nice to also add the max sliced wasserstein. Now that you have the new muodule most of the work is done indeed.

AdrienCorenflos commented:
Anything else @rflamary ?

rflamary commented:
Hello,

Thank you @AdrienCorenflos . I'm OK with the PR and plan on merging it.

But the build_linux_minimal fails and I want to investigate this problem before merging. It does not come from you PR but I don't wan't to merge if the build fails.

AdrienCorenflos commented:
Maybe you'd want to rerun the checks?

rflamary commented:
The problem is that if I rerun the checks they might pass but that something we have to understand and Im' pretty busy right now ;).

I promise I will merge your PR before the next release but I want to solve the mystery before pilling merges.
